### DISCUSSION TITLE
✨🏠 Building model

### DISCUSSION THREAD

**Discussion by BenWinchester** [CLOVER-energy/CLOVER#78] [Ideas]
# Building Model
## Introduction
The inclusion of HVAC loads in CLOVER requires, amongst other things.
* the inclusion of the notion of a `COOLING` `ResourceType`,
* information gathering as to the performance of various cooling technologies, to be stored in the `conversion_inputs.yaml` file,
* a new `building_inputs.yaml` file, which will contain information about heat-loss, occupancy, heat generation and ideal-temperature characteristics of the various buildings,
* a means of passing around cooling profiles and dealing with their conversion into electric/hot-water profiles.

## Building file
The Building File will need to contain thermal information about buildings which place heating/cooling demands on the system. The level of detail that CLOVER goes into regarding the modelling of buildings, e.g., whether buildings are modelled at the same level as hot-water tanks or whether they are modelled as having configurable walls etc., will affect the amount of information that goes into the building File.

The Building File will also contain information about the occupancy characteristics of the building, electrical devices employed etc. where the devices used are defined in more detail within the [Devices File](https://docs.google.com/document/d/1aT-4-7Kk35ivYbJ-bnv8Oj2bp65pUYdtHQLuKZtxN1o/edit#heading=h.ix1oql8g0sr8). The Building File will override the usage characteristics from the [Devices File](https://docs.google.com/document/d/1aT-4-7Kk35ivYbJ-bnv8Oj2bp65pUYdtHQLuKZtxN1o/edit#heading=h.ix1oql8g0sr8) with its own characteristics for the various devices included in the building. This is necessary in order for the heat produced by these devices to be correctly accounted for throughout the modelling period.

As such, each device which is used both by households and the specific buildings needs to contain:
* Usage profiles for community usage as well as use within the building,
* Associated heat-production characteristics so that the waste heat generated and added to the building can correctly be determined,
* A means of making it clear to the end user that if a device is “available” then it is available both within the community and within all the buildings which specify its use.
These usage profiles can be differentiated within the device-utilisation directory, either by the filename or by a device-based marker, with CLOVER reporting an error if a device is requested within the community and no profile is present, or is requested within a building and no profile is present.

The Building File will also contain information about the maximum and minimum temperatures acceptable/desired within the building which will be used to determine the HVAC load that the building places on the system.

## Building module
Depending on the time constant of the buildings considered, the complexity of the approach will vary. (More information is given in the [HVAC Building Loads](https://docs.google.com/document/d/1aT-4-7Kk35ivYbJ-bnv8Oj2bp65pUYdtHQLuKZtxN1o/edit#heading=h.u19ishsn4bfw) section.) For extensibility, the system is modelled dynamically with the load dependent on the previous temperature of the building.
**Note:** in the simplest case, the model will be steady-state, relying only on occupancy, device-usage and weather conditions to determine the cooling load.

### Repeatability Considerations
This shouldn’t result in any repeatability changes to CLOVER because:
* The load is calculated based on
  * the temperature profiles, which are generated/fetched once and which can be re-used in future simulations,
  * occupancy profiles, which will be generated once and can be used again in subsequent simulations and optimisations,
  * the weather profiles, which, again, are fetched once only per location.
* Whether or not the load was successfully supplied depends on PV-C / PV(-X) technological information, weather conditions, and grid-availability profiles which themselves depend on environmental, device-usage and occupancy profiles, again, all of which are static.
Hence, the same output should be generated provided the same temperature and supply profiles are used.

### Module Approach
The module will take in some previous temperature of the building, along with building properties and the ambient temperature/weather conditions, and will return the HVAC load required (and whether it is a heating or cooling load) such that the temperature within the building falls within the range specified in the [Building File](https://docs.google.com/document/d/1aT-4-7Kk35ivYbJ-bnv8Oj2bp65pUYdtHQLuKZtxN1o/edit#heading=h.pjy8uyfegkzh). This interaction is with the [Energy System Module](https://docs.google.com/document/d/1aT-4-7Kk35ivYbJ-bnv8Oj2bp65pUYdtHQLuKZtxN1o/edit#heading=h.tiw7vcsac2lo).
**Note:** in the simplest, _steady-state_ model, the previous temperature of the building does not affect the cooling load, as the calculation will be done based on whether some ideal-case demand is met.

Whether the HVAC load was successfully supplied can be fed back into the building in order to determine, based on environmental factors, the previous temperature of the building etc., the new temperature of the building which can be used at the next time step.
Building Class if a dynamic temperature is being used. At this point in the concept process, it is deemed likely that a steady-state model will be used.

***
For more information, contact the @CLOVER-energy/clover-development-team or consult the [PV-X in CLOVER: Design Specification](https://docs.google.com/document/d/1aT-4-7Kk35ivYbJ-bnv8Oj2bp65pUYdtHQLuKZtxN1o/edit?usp=sharing) document.

**Comment by BenWinchester** [comment]
Hi @IlariaDF ,

Hopefully some of the information above is useful? If not, it's at least useful for a future reader to understand this thread :smile:.

Some places where I think we'll need to change the code:
* [ResourceType](https://github.com/CLOVER-energy/CLOVER/blob/9aa0ef44aed90baf601dd898a384a47095f145b7/src/clover/__utils__.py#L1176) already has cooling as a concept, and the mapping used to parse the information in the inputs file has already got the [`cooling` keyword](https://github.com/CLOVER-energy/CLOVER/blob/9aa0ef44aed90baf601dd898a384a47095f145b7/src/clover/__utils__.py#L1231), so that's good :smile: ;
* We'll need a new buildings file, which should probably sit in the [simulation directory](https://github.com/CLOVER-energy/CLOVER/tree/ilaria/locations/Bahraich/inputs/simulation) as it's part of the energy system, though it could also sit within the load directory;
* The `total_loads` variable will need updating with a cooling load [here](https://github.com/CLOVER-energy/CLOVER/blob/9aa0ef44aed90baf601dd898a384a47095f145b7/src/clover/__main__.py#L801):
  ```diff
    # Assemble a means of storing the relevant loads.
    total_loads: Dict[ResourceType, Optional[pd.DataFrame]] = {
        ResourceType.CLEAN_WATER: total_cw_load,
  +     ResourceType.COOLING: total_cooling_load,
        ResourceType.ELECTRIC: 0.001 * total_electric_load,  # type: ignore
        ResourceType.HOT_CLEAN_WATER: total_hw_load,
    }
  ```
  which we'll need to calculate probably by calling our new building module. I don't think we'll be able to use the existing load module for the cooling load (though maybe we should be putting the code there) because we don't want to get the load by adding up devices etc., and, rather, want to look at weather conditions and process things in a different way.
  Basically, we want to follow the steps similar to/equal to that [calculation](https://theengineeringmindset.com/cooling-load-calculation-cold-room/) you found, rather than stochastically generating profiles... however... we probably do want to do some stochastic generation around occupancy etc. e.g., if the health centre has between 1 and 4 people in at any time, and it averages to 2, then we want some random noise to reflect this.
* We'll also need to put something new in the `energy_system.py` module, to point it at our new `buildings.py` module, but that's something later to do I think.

(We can add more things as we go :smile: )
